File: backend/app/schemas/review.py
"""Pydantic schemas for Review CRUD and AI pipeline status."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ReviewCreate(BaseModel):
    restaurant_id: int
    rating: int = Field(..., ge=1, le=5)
    comment: str | None = Field(None, max_length=2000)


class ReviewUpdate(BaseModel):
    rating: int | None = Field(None, ge=1, le=5)
    comment: str | None = Field(None, max_length=2000)


class ReviewResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    restaurant_id: int
    rating: int
    comment: str | None = None
    detected_language: str | None = None
    translated_text: str | None = None
    sentiment: str | None = None
    sentiment_score: float | None = None
    processing_status: str = "pending"
    created_at: datetime | None = None
    processed_at: datetime | None = None
    author_name: str | None = None
    author_username: str | None = None


class ReviewProcessingStatus(BaseModel):
    review_id: int
    processing_status: str
    detected_language: str | None = None
    sentiment: str | None = None
    sentiment_score: float | None = None
    processed_at: datetime | None = None

File: frontend/lib/models/review.dart
import 'json_parse.dart';

/// Restaurant review from `ReviewResponse`.
class Review {
  const Review({
    required this.id,
    required this.userId,
    required this.restaurantId,
    required this.rating,
    this.comment,
    this.sentiment,
    this.sentimentScore,
    this.authorName,
    this.authorUsername,
    this.createdAt,
  });

  final int id;
  final int userId;
  final int restaurantId;
  final int rating;
  final String? comment;
  final String? sentiment;
  final double? sentimentScore;
  final String? authorName;
  final String? authorUsername;
  final DateTime? createdAt;

  String get displayAuthor =>
      authorName?.trim().isNotEmpty == true
          ? authorName!
          : (authorUsername?.trim().isNotEmpty == true ? '@${authorUsername!}' : 'Guest');

  factory Review.fromJson(Map<String, dynamic> json) => Review(
        id: parseInt(json['id'], field: 'id'),
        userId: parseInt(json['user_id'], field: 'user_id'),
        restaurantId: parseInt(json['restaurant_id'], field: 'restaurant_id'),
        rating: parseInt(json['rating'], field: 'rating'),
        comment: json['comment']?.toString(),
        sentiment: json['sentiment']?.toString(),
        sentimentScore: parseDoubleOrNull(json['sentiment_score']),
        authorName: json['author_name']?.toString(),
        authorUsername: json['author_username']?.toString(),
        createdAt: parseDateTimeOrNull(json['created_at']),
      );
}
